src/libstd/thread/mod.rs

```diff
@@ -525,12 +526,73 @@ pub fn park_timeout(dur: Duration) {
 }
 
 ////////////////////////////////////////////////////////////////////////////////
+// ThreadId
+////////////////////////////////////////////////////////////////////////////////
+
+/// A unique identifier for a running thread.
+///
+/// A `ThreadId` is an opaque object that has a unique value for each thread
+/// that creates one. `ThreadId`s do not correspond to a thread's system-
+/// designated identifier.
+#[unstable(feature = "thread_id", issue = "21507")]
+#[derive(Eq, PartialEq, Copy, Clone)]
+pub struct ThreadId(u64);
+
+impl ThreadId {
+    // Generate a new unique thread ID. Since this function is called every
+    // time a thread is created, this is optimized to generate unique values
+    // as quickly as possible.
+    fn new() -> ThreadId {
+        // 64-bit operations are not atomic on all systems, so use an atomic
+        // flag as a guard around a 64-bit global counter. The window for
+        // contention on the counter is rather narrow since the general case
+        // should be compiled down to three instructions between locking and
+        // unlocking the guard. Since contention on the guard is low, use a
+        // spinlock that optimizes for the fast path of the guard being
+        // unlocked.
+        static GUARD: AtomicBool = AtomicBool::new(false);
+        static mut COUNTER: u64 = 0;
+
+        // Get exclusive access to the counter.
+        while GUARD.compare_exchange_weak(
+            false,
+            true,
+            Ordering::Acquire,
+            Ordering::Relaxed
+        ).is_err() {
+            // Give up the rest of our thread quantum if another thread is
+            // using the counter. This is the slow_er_ path.
+            yield_now();
```

Comment:
It's ok to in general not call `Mutex::destroy`, and otherwise I don't think there's going to be any real impact here on a mutex vs a spinlock. While `Once` uses atomics, it doesn't use spinlocks internally.

In general we try to avoid spinlocks wherever possible as there are generally more suitable synchronization primitives.